node/core/pvf/src/error.rs

```diff
@@ -78,13 +78,33 @@ pub enum InvalidCandidate {
 
 impl From<PrepareError> for ValidationError {
 	fn from(error: PrepareError) -> Self {
-		let error_str = match error {
-			PrepareError::Prevalidation(err) => format!("prevalidation: {}", err),
-			PrepareError::Preparation(err) => format!("preparation: {}", err),
-			PrepareError::Panic(err) => format!("panic: {}", err),
-			PrepareError::TimedOut => "preparation timeout".to_owned(),
-			PrepareError::DidNotMakeIt => "communication error".to_owned(),
-		};
-		ValidationError::InvalidCandidate(InvalidCandidate::WorkerReportedError(error_str))
+		// Here we need to classify the errors into two errors: deterministic and non-deterministic.
+		//
+		// Non-deterministic errors can happen spuriously. Typically, they occur due to resource
+		// starvation, e.g. under heavy load or memory pressure. Those errors are typically transient
+		// but may persist e.g. if the node is run by overwhelmingly underpowered machine.
+		//
+		// Deterministic errors should trigger reliably. Those errors depend on the PVF itself and
+		// the sc-executor/wasmtime logic.
+		//
+		// For now, at least until the PVF pre-checking lands, the deterministic errors will be
+		// treated as `InvalidCandidate`. Should those occur they could potentially trigger disputes.
+		//
+		// All non-deterministic errors are qualified as `InternalError`s and will not trigger
+		// disputes.
+		match error {
+			PrepareError::Prevalidation(err) => ValidationError::InvalidCandidate(
+				InvalidCandidate::PrepareError(format!("prevalidation: {}", err)),
```

Comment:
Can't we include a `PrepareError` instead of stringifing?

Reply:
That idea crossed my mind too and I ditched this:

1. We do not need to know (yet) what exactly was this error. Ultimately the string is required for supplying into `ExecutionError`.
2. We actually do not want the full `PrepareError` in `InvalidCandidate::PrepareError` because the full `PrepareError` contains both "deterministic" and "undeterminstic" error, whereas `InvalidCandidate::PrepareError` promises to pass only deterministic variants. While that's not the end of the world I prefer to not include it, because I strive to limit the impossible states on the type level and that also IMO could confuse people thinking that all those errors are possible.

Reply:
I see, I have an idea how to solve this by only allowing the sane, sparse set to occur, but that's for a follow up PR :)